File: src/dicepoker/game/ScoreTable.java
package dicepoker.game;

import java.util.EnumMap;
import java.util.Map;

import dicepoker.config.GameConfig;
import dicepoker.model.CombinationType;

/**
 * Таблица очков игрока.
 */
public class ScoreTable {
    private final Map<CombinationType, Integer> scores = new EnumMap<>(CombinationType.class);

    public boolean isComplete() {
        return scores.size() == GameConfig.AVAILABLE_COMBINATIONS.size();
    }

    public void record(CombinationType t, int s) {
        scores.put(t, s);
    }

    public boolean used(CombinationType t) {
        return scores.containsKey(t);
    }

    public int total() {
        return scores.values().stream().mapToInt(i -> i).sum() + bonus();
    }

    private int bonus() {
        int sum = 0;
        for (var t : CombinationType.values())
            if (t.name().matches("ONES|TWOS|THREES|FOURS|FIVES|SIXES"))
                sum += scores.getOrDefault(t, 0);
        return sum >= 63 ? 50 : 0;
    }

    public void printTable() {
        for (var t : GameConfig.AVAILABLE_COMBINATIONS) {
            String v = used(t) ? scores.get(t) + "" : "-";
            System.out.printf("%15s: %s\n", t, v);
        }
        System.out.println("Bonus: " + bonus());
        System.out.println("Total: " + total());
    }
}

File: dice_poker/score_table.py
from typing import Dict
from .combination_type import CombinationType


class ScoreTable:
    def __init__(self):
        self.scores: Dict[CombinationType, int] = {}

    def record(self, comb_type: CombinationType, score: int):
        self.scores[comb_type] = score

    def used(self, comb_type: CombinationType) -> bool:
        return comb_type in self.scores

    def is_complete(self) -> bool:
        return len(self.scores) == len(CombinationType)

    def total(self) -> int:
        base = sum(self.scores.values())
        upper = sum(self.scores.get(t, 0) for t in [
            CombinationType.ONES, CombinationType.TWOS, CombinationType.THREES,
            CombinationType.FOURS, CombinationType.FIVES, CombinationType.SIXES
        ])
        bonus = 50 if upper >= 63 else 0
        return base + bonus
